panic: Failed to extract orders as not enough volume within price limits

Running RL training from https://github.com/vegaprotocol/vega-market-sim/commit/145fbb0c9bb1e9813dd6f018c13b8b01e56b2c29 with 
```
python ./vega_sim/reinforcement/run_rl_agent.py --rl-max-it 200
```

This is what we fail with:
```
panic: Failed to extract orders as not enough volume within price limits

goroutine 5117594 [running]:
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0x1400e69a540, {0x1400d553200, 0x2, 0x2})
        /Users/davidsiska/go/pkg/mod/go.uber.org/zap@v1.21.0/zapcore/entry.go:232 +0x30c
go.uber.org/zap.(*Logger).Panic(0xe1dde?, {0x101b9fe27?, 0x0?}, {0x1400d553200, 0x2, 0x2})
        /Users/davidsiska/go/pkg/mod/go.uber.org/zap@v1.21.0/logger.go:230 +0x54
code.vegaprotocol.io/vega/core/matching.(*OrderBookSide).ExtractOrders(0x1400f686090, 0x140169dbfa0, 0x64, 0x0)
        /Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/matching/side.go:205 +0x600
code.vegaprotocol.io/vega/core/matching.(*OrderBook).GetIndicativeTrades(0x140113a6900)
```

The whole node error log:  [node.err.txt](https://github.com/vegaprotocol/vega/files/9708947/node.err.txt)


Last 20k lines from node output log: [node.out.tail-n20000.txt](https://github.com/vegaprotocol/vega/files/9708948/node.out.tail-n20000.txt)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -25,6 +25,7 @@
 - [6373](https://github.com/vegaprotocol/vega/issues/6373) - Fix panic with the metrics as well
 - [6395](https://github.com/vegaprotocol/vega/issues/6395) - Wallet selection doesn't lower case the wallet name during input verification
 - [6408](https://github.com/vegaprotocol/vega/issues/6408) - Initialise observer in liquidity provision `sql` store
+- [6406](https://github.com/vegaprotocol/vega/issues/6406) - Fix invalid tracking of cumulative volume and price
 
 ## 0.57.0
 

--- a/core/matching/cancelorders_test.go
+++ b/core/matching/cancelorders_test.go
@@ -88,8 +88,12 @@ func TestOrderBookSimple_CancelWrongOrderIncorrectMarketID(t *testing.T) {
 		Type:        types.OrderTypeLimit,
 		ID:          "v0000000000000-0000001",
 	}
-	_, err = book.CancelOrder(&order2)
-	assert.Error(t, err, types.ErrOrderRemovalFailure)
+
+	assert.Panics(t, func() {
+		_, err = book.CancelOrder(&order2)
+	},
+	)
+
 	assert.Equal(t, book.getNumberOfBuyLevels(), 1)
 	assert.Equal(t, book.getNumberOfSellLevels(), 0)
 }

--- a/core/matching/indicative_price_and_volume.go
+++ b/core/matching/indicative_price_and_volume.go
@@ -72,7 +72,7 @@ func NewIndicativePriceAndVolume(log *logging.Logger, buy, sell *OrderBookSide) 
 	return &ipv
 }
 
-// this will be user to build the initial set of price levels, when the auction is being started.
+// this will be used to build the initial set of price levels, when the auction is being started.
 func (ipv *IndicativePriceAndVolume) buildInitialCumulativeLevels(buy, sell *OrderBookSide) {
 	// we'll keep track of all the pl we encounter
 	mplm := map[num.Uint]ipvPriceLevel{}
@@ -271,6 +271,8 @@ func (ipv *IndicativePriceAndVolume) GetCumulativePriceLevels(maxPrice, minPrice
 
 	// reset those fields
 	ipv.needsUpdate = false
+	ipv.lastMinPrice = minPrice.Clone()
+	ipv.lastMaxPrice = maxPrice.Clone()
 	ipv.lastMaxTradable = maxTradable
 	ipv.lastCumulativeVolumes = cumulativeVolumes
 	return cumulativeVolumes, maxTradable

--- a/core/matching/orderbook.go
+++ b/core/matching/orderbook.go
@@ -610,12 +610,9 @@ func (b *OrderBook) CancelAllOrders(party string) ([]*types.OrderCancellationCon
 func (b *OrderBook) CancelOrder(order *types.Order) (*types.OrderCancellationConfirmation, error) {
 	// Validate Market
 	if order.MarketID != b.marketID {
-		if b.log.GetLevel() == logging.DebugLevel {
-			b.log.Debug("Market ID mismatch",
-				logging.Order(*order),
-				logging.String("order-book", b.marketID))
-		}
-		return nil, types.OrderErrorInvalidMarketID
+		b.log.Panic("Market ID mismatch",
+			logging.Order(*order),
+			logging.String("order-book", b.marketID))
 	}
 
 	// Validate Order ID must be present
@@ -763,10 +760,6 @@ func (b *OrderBook) SubmitOrder(order *types.Order) (*types.OrderConfirmation, e
 		b.latestTimestamp = order.CreatedAt
 	}
 
-	if b.LogPriceLevelsDebug {
-		b.PrintState("Entry state:")
-	}
-
 	var (
 		trades          []*types.Trade
 		impactedOrders  []*types.Order
@@ -781,10 +774,6 @@ func (b *OrderBook) SubmitOrder(order *types.Order) (*types.OrderConfirmation, e
 		if !lastTradedPrice.IsZero() {
 			b.lastTradedPrice = lastTradedPrice
 		}
-		// if state of the book changed show state
-		if b.LogPriceLevelsDebug && len(trades) != 0 {
-			b.PrintState("After uncross state:")
-		}
 	}
 
 	// if order is persistent type add to order book to the correct side
@@ -798,10 +787,6 @@ func (b *OrderBook) SubmitOrder(order *types.Order) (*types.OrderConfirmation, e
 		if b.auction {
 			b.indicativePriceAndVolume.AddVolumeAtPrice(
 				order.Price, order.Remaining, order.Side)
-		}
-
-		if b.LogPriceLevelsDebug {
-			b.PrintState("After addOrder state:")
 		}
 	}
 
@@ -882,13 +867,21 @@ func (b *OrderBook) DeleteOrder(
 		}
 		return nil, types.ErrOrderRemovalFailure
 	}
+
 	delete(b.ordersByID, order.ID)
 	delete(b.ordersPerParty[order.Party], order.ID)
 	delete(b.peggedOrders, order.ID)
-	// also add it to the indicative price and volume if in auction
+
+	// also remove it to the indicative price and volume if in auction
+	// here we use specifically the order which was in the book, just in case
+	// the order passed in would be wrong
+	// TODO: refactor this better, we should never need to pass in more that IDs there
+	// because by using the order passed in remain and price, we could use
+	// values which have been amended previously... (e.g: amending an order which
+	// cancel the order if it expires it
 	if b.auction {
 		b.indicativePriceAndVolume.RemoveVolumeAtPrice(
-			order.Price, order.Remaining, order.Side)
+			dorder.Price, dorder.Remaining, dorder.Side)
 	}
 	return dorder, err
 }

--- a/core/matching/orderbook_test.go
+++ b/core/matching/orderbook_test.go
@@ -1377,12 +1377,7 @@ func TestOrderBook_CancelOrderMarketMismatch(t *testing.T) {
 
 	orderAdded.MarketID = "invalid" // Bad market, malformed?
 
-	_, err = book.ob.CancelOrder(orderAdded)
-	if err != nil {
-		fmt.Println(err)
-	}
-
-	assert.Equal(t, types.OrderErrorInvalidMarketID, err)
+	assert.Panics(t, func() { _, err = book.ob.CancelOrder(orderAdded) })
 }
 
 func TestOrderBook_CancelOrderInvalidID(t *testing.T) {

--- a/core/matching/side.go
+++ b/core/matching/side.go
@@ -204,7 +204,10 @@ func (s *OrderBookSide) ExtractOrders(price *num.Uint, volume uint64, removeOrde
 				// We should never get to here unless the passed in price
 				// and volume are not correct
 				s.log.Panic("Failed to extract orders as not enough volume within price limits",
-					logging.BigUint("Price", price), logging.Uint64("volume", volume))
+					logging.BigUint("price", price),
+					logging.Uint64("required-volume", volume),
+					logging.Uint64("found-volume", totalVolume),
+					logging.Bool("remove-orders", removeOrders))
 			}
 
 			// If we have the right amount, stop processing

